utils/estimated_histogram.hh

```diff
@@ -47,13 +47,302 @@
+
+    /*
+     * \brief returns the count in the given bucket
+     */
+    uint64_t get(size_t bucket) const {
+        return _buckets[bucket];
+    }
+
+    /*!
+     * \brief get a histogram quantile
+     *
+     * returns the estimated value at given quantile
+     */
+    uint64_t quantile(double quantile) const {
+        if (quantile < 0 || quantile > 1.0) {
+            throw std::runtime_error("Invalid quantile value " + std::to_string(quantile) + ". Value should be between 0 and 1");
+        }
+        auto c = count();
+
+        if (!c) {
+            return 0; // no data
+        }
+
+        auto pcount = uint64_t(std::floor(c * quantile));
+        uint64_t elements = 0;
+        for (size_t i = 0; i < NumBuckets - 2; i++) {
+            if (_buckets[i]) {
+                elements += _buckets[i];
+                if (elements >= pcount) {
+                    return get_bucket_lower_limit(i);
+                }
+            }
+        }
+        return Max; // overflowed value is in the requested quantile
+    }
+
+    /*!
+     * \brief returns the mean histogram value (average of bucket offsets, weighted by count)
+     */
+    uint64_t mean() const {
+        double elements = 0;
+        uint64_t sum = 0;
+        for (size_t i = 0; i < NumBuckets - 1; i++) {
+            elements += _buckets[i];
+            sum += _buckets[i] * get_bucket_lower_limit(i);
+        }
+        return (sum + elements - 1) / elements;
+    }
+
+    /*!
+     * \brief returns the number of buckets;
+     */
+    size_t size() const {
+        return NumBuckets;
+    }
+
+    /*!
+     * \brief returns the total number of values inserted
+     */
+    uint64_t count() const {
+        uint64_t sum = 0L;
+        for (size_t i = 0; i < NumBuckets; i++) {
+            sum += _buckets[i];
+        }
+        return sum;
+    }
+
+    /*!
+     * \brief multiple all the buckets content in the histogram by a constant
+     */
+    approx_exponential_histogram& operator*=(double v) {
+        for (size_t i = 0; i < NumBuckets; i++) {
+            _buckets[i] *= v;
+        }
+        return *this;
+    }
+};
+
+template<uint64_t Min, uint64_t Max, size_t NumBuckets>
+inline approx_exponential_histogram<Min, Max, NumBuckets> base_estimated_histogram_merge(approx_exponential_histogram<Min, Max, NumBuckets> a, const approx_exponential_histogram<Min, Max, NumBuckets>& b) {
+    return a.merge(b);
+}
+
+/*!
+ * \brief estimated histogram for duration values
+ * time_estimated_histogram is used for common task timing.
+ * It covers the range of 0.5ms to 33s with a 2 bits granularity.
+ *
+ * 512us, 640us, 768us, 896us, 1024us, 1280us, 1536us, 1792us, 2048us, 2560us...
+ */
+class time_estimated_histogram : public approx_exponential_histogram<512, 33554432, 65> {
+public:
+    using clock = std::chrono::steady_clock;
+    using duration = clock::duration;
+    using approx_exponential_histogram<512, 33554432, 65>::add;
```

Comment:
Why do you need to do this "using", isn't it inherited by default?
By the way, below you provide different add_micro() and add() functions which I supposed you want users to use. I think it would have been better to hide the add(uint64_t) function, instead of exposing it. In add_micro() you could call it directly as approx_exponential_histogram::add().

Reply:
ok

Reply:
note that because of C++ overload, redefinition of add(duration) would hide add(uint64)

Reply:
Ok, I'm not sure why it hides, but I guess you checked it... In that case, even better that it hides, no? I thought the point of adding add_micro(...) was that the add(uint64_t) is too unclear, so it's better not to have it at all.
By the way, I would go even further, and hide *all* the functions of the base class (i.e., don't inherit public at all). Because all the mean(), quantile(), etc., could be overridden to return times instead of numbers.

Reply:
I believe they are useful especially for things like the API and the metrics layer.
I did follow your comment and hide the add(unit64) one